### ISSUE TITLE
Dummy Dataset for ASR purposes

### ISSUE THREAD

**Issue by jotix16** [rwth-i6/returnn#486]
For complicated `custom_construction_network` or MultiStage training such as in [returnn-experiments # 60](https://github.com/rwth-i6/returnn-experiments/pull/60) it would be nice to have some dummy dataset for fast testing. With testing, I mean checking if the network works as intended for different (sub)epochs.

One could of course use the dataset on hand, such as LibriSpeech, Switchboard or Timit but (sub)epochs are long and can take several minutes.

I have 2 ideas right now. 

One would be find out a way how to use only 1 sequence per (sub) epoch from the dataset on hand. That could be set as an option on the dataset class and hopefully not require too many changes. I tried looking into it but couldn't exactly find out how to do that. Maybe need to spend some more time with the `Dataset`.

One other way would be to have a `DummyDataset` class where we have full control and can decide `num_seq`, `input_seq_len`,`output_seq_len`, input_dim`, `output_dim` and the `Vocab`. 

Here I did some progress but am not able to properly incorporate the `Vocab`.

The dataset looks like this:
```python
class DummyDataset2(GeneratingDataset):
  """
  Some dummy data, which does not have any meaning.
  If you want to have artificial data with some meaning, look at other datasets here.
  The input are some dense data, the outputs are sparse.
  """

  def __init__(self, input_dim, output_dim, num_seqs, seq_len=20, output_seq_len=8,
               input_max_value=10.0, input_shift=None, input_scale=None, **kwargs):
    """
    :param int|None input_dim:
    :param int|dict[str,int|(int,int)|dict] output_dim:
    :param int|float num_seqs:
    :param int|dict[str,int] seq_len:
    :param float input_max_value:
    :param float|None input_shift:
    :param float|None input_scale:
    """
    super(DummyDataset2, self).__init__(input_dim=input_dim, output_dim=output_dim, num_seqs=num_seqs, **kwargs)
    self.seq_len = seq_len
    self.output_seq_len = output_seq_len
    self.input_max_value = input_max_value
    if input_shift is None:
      input_shift = -input_max_value / 2.0
    self.input_shift = input_shift
    if input_scale is None:
      input_scale = 1.0 / self.input_max_value
    self.input_scale = input_scale
    self.vocab = self.create_vocab(output_dim)

  def create_vocab(self, output_dim):
    import itertools
    chlist = 'ABCDEFG'
    base = len(chlist)
    power = 0
    tmp = output_dim
    while(tmp != 0):
      tmp = tmp // base
      power += 1
    assert base > power, "The choosen output_dim is too big"
    print(power)
    return [val for val in itertools.product(chlist, repeat=power)][:output_dim]

  def generate_seq(self, seq_idx):
    """
    :param int seq_idx:
    :rtype: DatasetSeq
    """
    seq_len = self.seq_len
    output_seq_len = self.output_seq_len
    i1 = seq_idx
    i2 = i1 + seq_len * self.num_inputs
    features = numpy.array([((i % self.input_max_value) + self.input_shift) * self.input_scale
                            for i in range(i1, i2)]).reshape((seq_len, self.num_inputs))
    i1, i2 = i2, i2 + output_seq_len
    targets = {"classes": numpy.array([i % self.num_outputs["classes"][0]
                                      for i in range(i1, i2)], dtype="int32")}
    targets["raw"] = numpy.array(["".join(self.vocab[t]) for t in targets["classes"]],
                                 dtype="object")
    # print("printing", targets)
    return DatasetSeq(
      targets=targets,
      features=features,
      seq_idx=seq_idx,
    )
```

Now when I use it in this way: 
```python
"""
DummyDataset in RETURNN automatically downloads the data via `nltk`,
so no preparation is needed.
This is useful for demos/tests.
"""

from __future__ import annotations
from typing import Dict, Any
from returnn.config import get_global_config
# from .librispeech.vocabs import bpe1k, bpe10k

from ..interface import DatasetConfig, VocabConfig


config = get_global_config()

# num_outputs = {'data': (40*2, 2), 'classes': (61, 1)}
# num_inputs = num_outputs["data"][0]
# _num_seqs = {'train': 144, 'dev': 16}


class DummyDataset(DatasetConfig):
  def __init__(self, audio_dim=50, output_dim=30, seq_len=30, output_seq_len=12, num_seqs=6, debug_mode=None):
    super(DummyDataset, self).__init__()
    if debug_mode is None:
      debug_mode = config.typed_dict.get("debug_mode", False)
    self.audio_dim = audio_dim
    self.output_dim = output_dim
    self.seq_len = seq_len
    self.output_seq_len = output_seq_len
    self.num_seqs = num_seqs
    self.debug_mode = debug_mode

  def get_extern_data(self) -> Dict[str, Dict[str, Any]]:
    return {
      "data": {"sparse": True, "dim": self.audio_dim},
      "classes": {"sparse": True, "dim": self.output_dim},
    }

  def get_train_dataset(self) -> Dict[str, Any]:
    return self.get_dataset("train")

  def get_eval_datasets(self) -> Dict[str, Dict[str, Any]]:
    return {
      "dev": self.get_dataset("dev"),
      "devtrain": self.get_dataset("train")}

  def get_dataset(self, key, subset=None):
    assert key in {"train", "dev"}
    print(f"Using {key} dataset!")
    return {
      "class": "DummyDataset2",
      "input_dim": self.audio_dim,
      "output_dim": self.output_dim,
      "seq_len": self.seq_len,
      "output_seq_len": self.output_seq_len,
      "num_seqs": self.num_seqs,
      # "input_base": 10
    }
```

the `dump-dataset.py` function doesn't output the `raw`/text part of the targets:

``` bash
Train data:
  input: 50 x 1
  output: {'classes': (30, 1), 'data': (50, 2)}
  DummyDataset2, sequences: 6, frames: unknown
Epoch: 1
Dataset keys: ['data', 'classes']
Dataset target keys: ['classes']
Dump to stdout
seq 0/6 (16.67%) (0:00:00) tag: seq-0
seq 0/6 (16.67%) (0:00:00) data: array([[-0.5, -0.4, -0.3, ...,  0.2,  0.3,  0.4],
       [-0.5, -0.4, -0.3, ...,  0.2,  0.3,  0.4],
       [-0.5, -0.4, -0.3, ...,  0.2,  0.3,  0.4],
       ...,
       [-0.5, -0.4, -0.3, ...,  0.2,  0.3,  0.4],
       [-0.5, -0.4, -0.3, ...,  0.2,  0.3,  0.4],
       [-0.5, -0.4, -0.3, ...,  0.2..., shape=(30, 50)
seq 0 target 'classes': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int32), shape=(12,)
seq 1/6 (33.33%) (0:00:00) tag: seq-1
seq 1/6 (33.33%) (0:00:00) data: array([[-0.4, -0.3, -0.2, ...,  0.3,  0.4, -0.5],
       [-0.4, -0.3, -0.2, ...,  0.3,  0.4, -0.5],
       [-0.4, -0.3, -0.2, ...,  0.3,  0.4, -0.5],
       ...,
       [-0.4, -0.3, -0.2, ...,  0.3,  0.4, -0.5],
       [-0.4, -0.3, -0.2, ...,  0.3,  0.4, -0.5],
       [-0.4, -0.3, -0.2, ...,  0.3..., shape=(30, 50)
seq 1 target 'classes': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int32), shape=(12,)
seq 2/6 (50.00%) (0:00:00) tag: seq-2
seq 2/6 (50.00%) (0:00:00) data: array([[-0.3, -0.2, -0.1, ...,  0.4, -0.5, -0.4],
       [-0.3, -0.2, -0.1, ...,  0.4, -0.5, -0.4],
       [-0.3, -0.2, -0.1, ...,  0.4, -0.5, -0.4],
       ...,
       [-0.3, -0.2, -0.1, ...,  0.4, -0.5, -0.4],
       [-0.3, -0.2, -0.1, ...,  0.4, -0.5, -0.4],
       [-0.3, -0.2, -0.1, ...,  0.4..., shape=(30, 50)
seq 2 target 'classes': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13], dtype=int32), shape=(12,)
seq 3/6 (66.67%) (0:00:00) tag: seq-3
seq 3/6 (66.67%) (0:00:00) data: array([[-0.2, -0.1,  0. , ..., -0.5, -0.4, -0.3],
       [-0.2, -0.1,  0. , ..., -0.5, -0.4, -0.3],
       [-0.2, -0.1,  0. , ..., -0.5, -0.4, -0.3],
       ...,
       [-0.2, -0.1,  0. , ..., -0.5, -0.4, -0.3],
       [-0.2, -0.1,  0. , ..., -0.5, -0.4, -0.3],
       [-0.2, -0.1,  0. , ..., -0.5..., shape=(30, 50)
seq 3 target 'classes': array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14], dtype=int32), shape=(12,)
seq 4/6 (83.33%) (0:00:00) tag: seq-4
seq 4/6 (83.33%) (0:00:00) data: array([[-0.1,  0. ,  0.1, ..., -0.4, -0.3, -0.2],
       [-0.1,  0. ,  0.1, ..., -0.4, -0.3, -0.2],
       [-0.1,  0. ,  0.1, ..., -0.4, -0.3, -0.2],
       ...,
       [-0.1,  0. ,  0.1, ..., -0.4, -0.3, -0.2],
       [-0.1,  0. ,  0.1, ..., -0.4, -0.3, -0.2],
       [-0.1,  0. ,  0.1, ..., -0.4..., shape=(30, 50)
seq 4 target 'classes': array([ 4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15], dtype=int32), shape=(12,)
seq 5/6 (100.00%) (0:00:00) tag: seq-5
seq 5/6 (100.00%) (0:00:00) data: array([[ 0. ,  0.1,  0.2, ..., -0.3, -0.2, -0.1],
       [ 0. ,  0.1,  0.2, ..., -0.3, -0.2, -0.1],
       [ 0. ,  0.1,  0.2, ..., -0.3, -0.2, -0.1],
       ...,
       [ 0. ,  0.1,  0.2, ..., -0.3, -0.2, -0.1],
       [ 0. ,  0.1,  0.2, ..., -0.3, -0.2, -0.1],
       [ 0. ,  0.1,  0.2, ..., -0.3..., shape=(30, 50)
seq 5 target 'classes': array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16], dtype=int32), shape=(12,)
Done. Total time 0:00:00.0250. More seqs which we did not dumped: False
Seq-length 'data' Stats:
  6 seqs
  Mean: 30.0
  Std dev: 0.0
  Min/max: 30 / 30
Seq-length 'classes' Stats:
  6 seqs
  Mean: 12.0
  Std dev: 0.0
  Min/max: 12 / 12
```

**Comment by albertz** [comment]
I think also you want to have a separate tool like `tools/test-network.py` or so?

The dummy dataset would just be created such that it matches `extern_data` (or `num_inputs`/`num_outputs` for old configs).

**Comment by jotix16** [comment]
The dummy dataset above is the same as `DummyDatasetMultipleSequenceLength`, only that I added a `Vocab` for the target labels.

> I think also you want to have a separate tool like tools/test-network.py or so?

That would be the cleanest solution actually. `tools/test-network.py` should look as similar as possible to the normal training only that the dataset is much smaller(one sequence per subepoch). And we can print some more information on the way too.

**Comment by albertz** [comment]
> The dummy dataset above is the same as `DummyDatasetMultipleSequenceLength`, only that I added a `Vocab` for the target labels.

Why do you need the labels (vocab) in the dataset for testing?

And if it is the same otherwise, why not extend `DummyDatasetMultipleSequenceLength` instead?

> > I think also you want to have a separate tool like tools/test-network.py or so?
> 
> That would be the cleanest solution actually.

Yes, I thought that was the intention. But in any case, such tool would then internally make use of such a dataset.
(But I think `DummyDatasetMultipleSequenceLength` or maybe also `StaticDataset` are already fine.)

> `tools/test-network.py` should look as similar as possible to the normal training 

Yes, you would probably just call `test-network.py <config> <other-args>` instead of `rnn.py <config> <other-args>`.

> only that the dataset is much smaller(one sequence per subepoch).

It should be a couple of seqs, such that batch size is >1 (with bs 1, you sometimes can hide some bugs which occur only with bs >1). And also you want that it runs for more than one step per (sub)epoch, because some bugs might only be triggered in step 2 or later (e.g. keep-over-epoch logic of hidden state or so).

> And we can print some more information on the way too.

Like what?
